developer/developer-guide.md

```diff
@@ -1,3 +1,872 @@
+and properties separately.
+
+## Data objects
+Alfred API has data objects that mirror the Alfresco concepts of QName, NodeRef, StoreRef, Path, 
+ContentData and ContentInputStream. These data objects are used to communicate with the 
+Alfred Java API without being dependent on Alfresco data types.
+
+Conversion between Alfresco and Alfred API data objects is the responsibility of the 
+`ApixToAlfrescoConversion` service. It is also possible to construct an Alfred API data object by
+passing its string representation to the constructor.
+
+## REST API
+The Alfred REST API is a thin wrapper around the Java abstraction layer. It converts its received
+parameters to the corresponding Alfred API data objects, then calls the corresponding service and
+serializes its return value to JSON.
+
+### Developer notes on search syntax
+#### Unimplemented search options:
+These options have some code towards handling, but are not implemented such that they are used in the search.
+* `facets.mincount`
+* `facets.limit`
+* `orderBy.expression`
+
+## Services
+Only the most important services are described here. Full documentation is available in 
+[the generated JavaDoc](#viewing-javadoc).
+
+### NodeService
+The `NodeService` provides operations on nodes.
+
+* Fetch and modify the metadata for a Node
+* Fetch the root node of a Store
+* Fetch, create and remove child, parent and target associations for a Node
+* Copy or move a Node to another parent
+* Create and delete Nodes
+* Checkout, checkin and fetch working copies for a Node
+
+### SearchService
+The `SearchService` allows searching for nodes based on an object tree.
+
+The `SearchService.query()` method takes a `SearchQuery` object which contains the search query to execute,
+as well as pagination, faceting and ordering options.
+
+\define{EXAMPLE_IMPORTS}
+\define{EXAMPLE_SEARCH_QUERY_OPTS}
+```java
+\include{examples/src/main/java/searchQuery.java}
+```
+\undef{EXAMPLE_IMPORTS}
+\undef{EXAMPLE_SEARCH_QUERY_OPTS}
+
+The query itself can be constructed using the `QueryBuilder`, which provides a fluent interface to 
+build search queries.
+
+\define{EXAMPLE_SEARCH_QUERY_QUERY}
+```java
+\include{examples/src/main/java/searchQuery.java}
+```
+\undef{EXAMPLE_SEARCH_QUERY_QUERY}
+
+When using the REST API, a JSON payload describing the search query has to be POST'ed to the 
+`apix/v1/search` endpoint. 
+This JSON document reflects the node structure created by the query builder, and is shown below:
+
+```json
+\include{examples/src/main/resources/jsonsearchquery.json}
+```
+
+### DictionaryService
+The `DictionaryService` provides meta-information about the metadata model.
+It allows to fetch information about registered types, aspect and properties.
+
+### Viewing JavaDoc
+Full JavaDoc documentation of the Alfred API Java interface is available in the JavaDoc. You can
+view the JavaDoc by browsing to `/alfresco/s/apix/javadocs/index.html` on your Alfresco host that
```

Comment:
I just tried this link https://demo.xenit.eu/alfresco/s/apix/javadocs/index.html and it gives 404. That's a problem that's out of scope for this ticket though. I made a ticket for it https://xenitsupport.jira.com/browse/ALFREDAPI-464